File: libReQL/Connection.m
/*
Copyright 2015 Adam Grandquist

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
 */
/**
 * @author Adam Grandquist
 * @copyright Apache
 */

#import "Connection.h"

#import "Query.h"

#import "./reql/core.h"

@implementation ReQLConnection {
  ReQL_Conn_t *p_conn;
}

-(instancetype)init {
  if (self = [super init]) {
    p_conn = malloc(sizeof(ReQL_Conn_t));
    if (p_conn == NULL) {
      [self release];
      return NULL;
    }
    reql_conn_init(p_conn);
    ReQL_Byte buf[500];
    if (reql_conn_connect(p_conn, buf, 500) != 0) {
      [self release];
      return NULL;
    }
  }
  return self;
}

-(BOOL)isOpen {
  return reql_conn_open(p_conn) == 0 ? NO : YES;
}

-(void)close {
  reql_conn_close(p_conn);
}

-(void)dealloc {
  reql_conn_destroy(p_conn);
  free(p_conn);
  [super dealloc];
}

@end

File: src/Lua/connection.cpp
/*
Copyright 2015 Adam Grandquist

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
 */
/**
 * @author Adam Grandquist
 * @copyright Apache
 */

#include "./Lua/connection.hpp"

#include "./reql/core.hpp"

#include <lua.h>
#include <lauxlib.h>
#include <lualib.h>

extern int
reql_lua_connect(lua_State *L) {
  lua_settop(L, 3);

  ReQL_Conn_t *conn = lua_newuserdata(L, sizeof(ReQL_Conn_t));

  reql_conn_init(conn);
  ReQL_Byte msg[500];

  if (reql_conn_connect(conn, msg, 500) != 0) {
    lua_pop(L, 1);
    return 0;
  }

  return 1;
}
